pkg/operator/ceph/webhook.go

```diff
@@ -18,58 +18,130 @@ package operator
 
 import (
 	"context"
-	"fmt"
 	"io/ioutil"
 	"os"
 	"path"
 
+	cs "github.com/jetstack/cert-manager/pkg/client/clientset/versioned/typed/certmanager/v1"
 	"github.com/pkg/errors"
 	"github.com/rook/rook/pkg/clusterd"
 	"github.com/rook/rook/pkg/operator/k8sutil"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/version"
+	apiadmissionsclient "k8s.io/client-go/kubernetes/typed/admissionregistration/v1"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
 const (
 	admissionControllerAppName       = "rook-ceph-admission-controller"
 	tlsPort                    int32 = 443
+	minK8sVersionWebhook             = "1.16.0"
+	webhookEnv                       = "ROOK_DISABLE_ADMISSION_CONTROLLER"
 )
 
 var (
-	namespace = os.Getenv(k8sutil.PodNamespaceEnvVar)
+	namespace              = os.Getenv(k8sutil.PodNamespaceEnvVar)
+	certManagerWebhookName = "cert-manager-webhook"
 )
 
-func isSecretPresent(ctx context.Context, context *clusterd.Context) (bool, error) {
-	logger.Infof("looking for admission webhook secret %q", admissionControllerAppName)
-	s, err := context.Clientset.CoreV1().Secrets(namespace).Get(ctx, admissionControllerAppName, metav1.GetOptions{})
+type WebhookContext struct {
+	// CertManager webhook
+	Admission cs.CertmanagerV1Interface
+
+	// Webhook
+	Webhook apiadmissionsclient.AdmissionregistrationV1Interface
+}
+
+func createWebhook(ctx context.Context, context *clusterd.Context) (bool, error) {
+	k8sVersion, err := k8sutil.GetK8SVersion(context.Clientset)
 	if err != nil {
-		// If secret is not found. All good ! Proceed with rook without admission controllers
-		if apierrors.IsNotFound(err) {
-			logger.Infof("admission webhook secret %q not found. proceeding without the admission controller", admissionControllerAppName)
+		logger.Errorf("failed to get k8s version. %v", err)
+		return false, nil
+	}
+	if !k8sVersion.AtLeast(version.MustParseSemantic(minK8sVersionWebhook)) {
+		logger.Errorf("minimum versions %s required for webhook to run", err)
+		return false, nil
+	}
+
+	var webhookCtx WebhookContext
```

Comment:
I think this is the way I see we do for all other clients and also in other projects also this is the way it is written. Let's discuss

Reply:
We create context structs in other places because they have a longer lifetime of broader scope where the context is needed. But since this is just a small scope of code when the operator is starting, the context struct isn't really needed. Let's discuss, not a big deal though.

Reply:
make sense, let's remove it.